lms/djangoapps/courseware/masquerade.py

```diff
@@ -116,3 +174,146 @@ def get_masquerading_group_info(user, course_key):
+        in that dict.
+
+        Arguments:
+            kv_dict (dict): dict mapping from `DjangoKeyValueStore.Key`s to field values
+        """
+        for key, value in kv_dict.iteritems():
+            self.session_data[self._serialize_key(key)] = value
+            self.session.modified = True
+
+    @contract(key=DjangoKeyValueStore.Key)
+    def delete(self, key):
+        """
+        Delete the value specified by `key`.
+
+        Arguments:
+            key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Raises: KeyError if key isn't found in the cache
+        """
+        # We can't simply delete the key from the session, since it might still exist in the cache,
+        # which we are not allowed to modify, so we mark it as deleted by setting it to
+        # _deleted_sentinel in the session.
+        self.session_data[self._serialize_key(key)] = _deleted_sentinel
+        self.session.modified = True
+
+    @contract(key=DjangoKeyValueStore.Key, returns=bool)
+    def has(self, key):
+        """
+        Return whether the specified `key` is set.
+
+        Arguments:
+            key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: bool
+        """
+        try:
+            value = self.session_data[self._serialize_key(key)]
+        except KeyError:
+            return self.cache.has(key)
+        else:
+            return value != _deleted_sentinel
+
+    # @contract(user_id=int, usage_key=UsageKey, score="number|None", max_score="number|None")
+    def set_score(self, user_id, usage_key, score, max_score):
+        """
+        Do nothing, since we are masquerading as a different user, so we don't want to set the score.
+        """
+        pass
+
+    @contract(key=DjangoKeyValueStore.Key, returns="datetime|None")
+    def last_modified(self, key):
+        """
+        Return when the supplied field was changed.
+
+        Arguments:
+            key (`DjangoKeyValueStore.Key`): The field value to look up
+
+        Returns: datetime if there was a modified date, or None otherwise
+        """
+        if self._serialize_key(key) in self.session_data:
+            # We don't keep track of modification times for data stored in the session.
+            return None
+        return self.cache.last_modified(key)
+
+
+def filter_displayed_blocks(block, view, frag, context):
+    """
+    A wrapper to only show XBlocks that set `show_in_read_only_mode` when masquerading as a specific user.
+
+    We don't want to modify the state of the user we are masquerading as, so we can't show XBlocks
+    that store information outside of the XBlock fields API.
+    """
+    if getattr(block, 'show_in_read_only_mode', False):
+        return frag
+    return Fragment(
+        u'Block of type {class_name} not whitelisted for display while masquerading as a'
+        u'specific student is active.'.format(class_name=block.unmixed_class.__name__)
```

Comment:
That wording is certainly better.  As a non-native speaker, I didn't spend much time thinking about a good message, since I learned there are much better qualified people from edX looking over the messages anyway.